warehouse/utils/wsgi.py

```diff
@@ -104,8 +132,31 @@ def _remote_addr_hashed(request: Request) -> str:
     return request.environ.get("REMOTE_ADDR_HASHED", "")
 
 
+def _ip_address(request):
+    """Return the IpAddress object for the remote address from the environment."""
+    try:
+        ip_address = (
+            request.db.query(IpAddress)
+            .filter(IpAddress.ip_address == request.remote_addr)
+            .one()
+        )
+    except NoResultFound:
+        ip_address = IpAddress(ip_address=request.remote_addr)
+        request.db.add(ip_address)
+
+    ip_address.hashed_ip_address = request.remote_addr_hashed
+    ip_address.geoip_info = {
+        k: request.environ[f"GEOIP_{v}"]
+        for k, v in GEOIP_FIELDS.items()
+        if f"GEOIP_{v}" in request.environ
+    }
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**question:** ‏How and when do these values get persisted to the db?

Reply:
at the end of the request, if everything was successful. same story as last time you asked, the transaction manager picks up things added to the session (see line 145).